zkevm-circuits/src/tx_circuit/sign_verify.rs

```diff
@@ -0,0 +1,978 @@
+                let offset = &mut 0;
+                let mut ctx = RegionCtx::new(&mut region, offset);
+                for i in 0..MAX_VERIF {
+                    let tx = if i < txs.len() {
+                        txs[i].clone()
+                    } else {
+                        // padding (enabled when address == 0)
+                        SignData::default()
+                    };
+                    let assigned_ecdsa = self.assign_ecdsa(&mut ctx, &chips, &tx)?;
+                    assigned_ecdsas.push(assigned_ecdsa);
+                }
+                Ok(())
+            },
+        )?;
+
+        let mut assigned_sig_verifs = Vec::new();
+        layouter.assign_region(
+            || "signature address verify",
+            |mut region| {
+                assigned_sig_verifs.clear();
+                // for i in 0..MAX_VERIF
+                for (i, assigned_ecdsa) in assigned_ecdsas.iter().enumerate() {
+                    let sign_data = txs.get(i); // None when padding (enabled when address == 0)
+                    let offset = i;
+                    let (assigned_sig_verif, keccak_aux) = self.assign_signature_verify(
+                        config,
+                        &mut region,
+                        offset,
+                        randomness,
+                        &address_is_zero_chip,
+                        sign_data,
+                        assigned_ecdsa,
+                    )?;
+                    if i < txs.len() {
+                        keccak_auxs.push(keccak_aux);
+                    }
+                    assigned_sig_verifs.push(assigned_sig_verif);
+                }
+
+                Ok(())
+            },
+        )?;
+
+        config.load_keccak(layouter, keccak_auxs, randomness)?;
+        config.load_range(layouter)?;
+
+        Ok(assigned_sig_verifs)
+    }
+}
+
+#[derive(Clone, Debug)]
+pub(crate) struct SignData {
+    pub(crate) signature: (secp256k1::Fq, secp256k1::Fq),
+    pub(crate) pk: Secp256k1Affine,
+    pub(crate) msg_hash: secp256k1::Fq,
+}
+
+// Returns (r, s)
+fn sign(
+    randomness: secp256k1::Fq,
+    sk: secp256k1::Fq,
+    msg_hash: secp256k1::Fq,
+) -> (secp256k1::Fq, secp256k1::Fq) {
+    let randomness_inv =
+        Option::<secp256k1::Fq>::from(randomness.invert()).expect("cannot invert randomness");
+    let generator = Secp256k1Affine::generator();
+    let sig_point = generator * randomness;
+    let x = *Option::<Coordinates<_>>::from(sig_point.to_affine().coordinates())
+        .expect("point is the identity")
+        .x();
+
+    let x_repr = &mut Vec::with_capacity(32);
+    x.write(x_repr).expect("cannot write bytes to array");
+
+    let mut x_bytes = [0u8; 64];
+    x_bytes[..32].copy_from_slice(&x_repr[..]);
+
+    let sig_r = secp256k1::Fq::from_bytes_wide(&x_bytes); // get x cordinate (E::Base) on E::Scalar
+    let sig_s = randomness_inv * (msg_hash + sig_r * sk);
+    (sig_r, sig_s)
+}
+
+lazy_static! {
```

Comment:
I would add here `//TODO: turn them into const initializations`

Reply:
I agree that it would be nice to have a const initialization here, but in order to get there we would need changes in a few external crates (because this initialization calls many functions from external crates that are not const).  So I think this would be a TODO that would never get done.

Reply:
+1